packages/elements/src/components/test-file/test-file.component.ts

```diff
@@ -76,27 +108,52 @@ export class MutationTestReportTestFile extends LitElement {
 
   private renderCode() {
     if (this.model?.source) {
-      return html`<pre id="report-code-block" class="line-numbers"><code class="language-${determineLanguage(this.model.name)}"><span>${unsafeHTML(
-        markTests(this.model.source, this.model.tests)
-      )}</span></code></pre>`;
+      const testsByLine = new Map<number, TestModel[]>();
+      for (const test of this.tests) {
+        if (test.location) {
+          let tests = testsByLine.get(test.location.start.line);
+          if (!tests) {
+            tests = [];
+            testsByLine.set(test.location.start.line, tests);
+          }
+          tests.push(test);
+        }
+      }
+
+      const renderFinalTests = (lastLine: number) => {
+        return this.renderTests([...testsByLine.entries()].filter(([line]) => line > lastLine).flatMap(([, tests]) => tests));
+      };
+
+      return html`<pre id="report-code-block" class="line-numbers"><code class="language-${determineLanguage(this.model.name)}"><table>
+        ${this.lines.map(
+        (line, lineNr) =>
+          html`<tr class="line"
+            ><td class="line-number"></td><td class="line-marker"></td
+            ><td class="code"
+              >${unsafeHTML(line)}${this.renderTests(testsByLine.get(lineNr + 1))}${this.lines.length === lineNr + 1
+                ? renderFinalTests(lineNr + 1)
+                : ''}</td
+            ></tr
+          >`
+      )}</table></code></pre>`;
     }
     return;
   }
 
-  private highlightCode() {
-    const code = this.shadowRoot?.querySelector('code');
-    if (code) {
-      highlightElement(code);
-
-      // Prism-js's `highlightElement` creates a copy of the DOM tree to do its magic.
-      // Now that the code is highlighted, we can bind the test components
-      this.forEachTestComponent((testComponent) => {
-        testComponent.test = this.model?.tests.find((test) => test.id === testComponent.getAttribute('test-id'));
-      });
-    }
+  private renderTests(tests: TestModel[] | undefined) {
```

Comment:
I think this method should be named differently, it's not rendering tests, it's rendering dots that mark a test. And maybe it should also be in a separate component (to be shared with source file component?)

Reply:
A separate component is a nice improvement, especially since we can then share it between mutant-dots and test-dots. For now, I'll just rename it if that's ok.